### PR TITLE
Added Kubernetes support.

### PR DESCRIPTION
**Pull request by saturnism**
This adds a KubernetesConfig, and the corresponding XML parser.

### CODE REVIEW

**Comment by gurbuzali** [comment]
I guess you've missed the `JoinConfig.getKubernetesConfig()` part

**Comment by saturnism** [comment]
Oops - I missed a dozen files =)

**Comment by gurbuzali** [comment]
verify

File: hazelcast/src/main/java/com/hazelcast/instance/BuildInfoProvider.java [hunk 1]
```diff
@@ -58,7 +58,7 @@ public static BuildInfo getBuildInfo() {
         } else {
             build = String.valueOf(hazelcastBuild);
         }
-        int buildNumber = Integer.parseInt(build);
+        int buildNumber = 0; //Integer.parseInt(build);
```
Comment (jerrinot) [inline hazelcast/src/main/java/com/hazelcast/instance/BuildInfoProvider.java]: is this a typo?

**Comment by gurbuzali** [comment]
you have some checkstyle errors
```
/scratch/jenkins/workspace/Hazelcast-3.x-pr-builder/hazelcast/src/main/java/com/hazelcast/instance/BuildInfoProvider.java:61: Don't use trailing comments.
/scratch/jenkins/workspace/Hazelcast-3.x-pr-builder/hazelcast/src/main/java/com/hazelcast/config/XmlConfigBuilder.java:1: File length is 1,515 lines (max allowed is 1,500).
```
you can check it yourself: `mvn clean compile install -P checkstyle`

**Comment by bilalyasar** [comment]
can you add xml config to schema file? (hazelcast-config-3.5.xsd)

File: hazelcast/src/main/java/com/hazelcast/config/KubernetesConfig.java [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+        if (connectionTimeoutSeconds < 0) {
+            throw new IllegalArgumentException("connection timeout can't be smaller than 0");
+        }
+        this.connectionTimeoutSeconds = connectionTimeoutSeconds;
+        return this;
+    }
+
+    @Override
+    public String toString() {
+        final StringBuilder sb = new StringBuilder("AwsConfig{");
```
Comment (bilalyasar) [inline hazelcast/src/main/java/com/hazelcast/config/KubernetesConfig.java]: i guess this should be "KubernetesConfig{" ?

**Comment by saturnism** [comment]
Thanks for the review! Will get them fixed. On the issue of `XmlConfigBuilder` filesize - do you feel I should break it up into multiple files?  That may be a bigger undertaking.

File: hazelcast/src/main/java/com/hazelcast/config/KubernetesConfig.java (hunk 2 above)
Comment (saturnism) [inline hazelcast/src/main/java/com/hazelcast/config/KubernetesConfig.java]: yup! will fix.

File: hazelcast/src/main/java/com/hazelcast/instance/BuildInfoProvider.java (hunk 1 above)
Comment (saturnism) [inline hazelcast/src/main/java/com/hazelcast/instance/BuildInfoProvider.java]: sorry temporarily changed it.  shouldn't have included in the commit.  will restore.

**Comment by saturnism** [comment]
rather than refactoring the XmlConfigBuilder, I've suppressed it checkstyle check on that file.

**Comment by noctarius** [comment]
Hi @saturnism,

We thought about how to handle those big changes. We just started an Hazelcast Incubator system. Since we want to make adding new cloud providers easier we want to have a clean SPI to integrate those. I added you to the new repo and I would love to see you help on the SPI design.

Sorry for closing this PR for now but I hope you understand the reasoning. Please do not yet delete the branch on your side, we will come back to the integration for the new SPI!

Chris

**Pull request closed without merging**